drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c

```diff
@@ -0,0 +1,1669 @@
+	port = get_port_data(node);
+	if (!port)
+		return 0;
+
+	setup_mmal_port_format(node, port);
+	ret = vchiq_mmal_port_set_format(dev->mmal_instance, port);
+	if (ret) {
+		v4l2_err(&dev->v4l2_dev,
+			 "%s: Failed vchiq_mmal_port_set_format on port, ret %d\n",
+			 __func__, ret);
+		ret = -EINVAL;
+	}
+
+	if (q_data->sizeimage < port->minimum_buffer.size) {
+		v4l2_err(&dev->v4l2_dev,
+			 "%s: Current buffer size of %u < min buf size %u - driver mismatch to MMAL\n",
+			 __func__,
+			 q_data->sizeimage,
+			 port->minimum_buffer.size);
+	}
+
+	v4l2_dbg(1, debug, &dev->v4l2_dev,
+		 "%s: Set format for type %d, wxh: %dx%d, fmt: %08x, size %u\n",
+		 __func__, f->type, q_data->width, q_data->height,
+		 q_data->fmt->fourcc, q_data->sizeimage);
+
+	return ret;
+}
+
+static int bcm2835_isp_node_querycap(struct file *file, void *priv,
+				     struct v4l2_capability *cap)
+{
+	strscpy(cap->driver, BCM2835_ISP_NAME, sizeof(cap->driver));
+	strscpy(cap->card, BCM2835_ISP_NAME, sizeof(cap->card));
+	snprintf(cap->bus_info, sizeof(cap->bus_info), "platform:%s",
+		 BCM2835_ISP_NAME);
+
+	return 0;
+}
+
+static int bcm2835_isp_node_g_fmt(struct file *file, void *priv,
+				  struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+
+	if (f->type != node->queue.type)
+		return -EINVAL;
+
+	if (node_is_stats(node)) {
+		f->fmt.meta.dataformat = V4L2_META_FMT_BCM2835_ISP_STATS;
+		f->fmt.meta.buffersize =
+			get_port_data(node)->minimum_buffer.size;
+	} else {
+		struct bcm2835_isp_q_data *q_data = &node->q_data;
+
+		f->fmt.pix.width = q_data->width;
+		f->fmt.pix.height = q_data->height;
+		f->fmt.pix.field = V4L2_FIELD_NONE;
+		f->fmt.pix.pixelformat = q_data->fmt->fourcc;
+		f->fmt.pix.bytesperline = q_data->bytesperline;
+		f->fmt.pix.sizeimage = q_data->sizeimage;
+		f->fmt.pix.colorspace = q_data->fmt->colorspace;
+	}
+
+	return 0;
+}
+
+static int bcm2835_isp_node_enum_fmt(struct file *file, void  *priv,
+				     struct v4l2_fmtdesc *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_fmt_list *fmts = &node->supported_fmts;
+
+	if (f->type != node->queue.type)
+		return -EINVAL;
+
+	if (f->index < fmts->num_entries) {
+		/* Format found */
+		f->pixelformat = fmts->list[f->index].fourcc;
+		f->flags = fmts->list[f->index].flags;
+		return 0;
+	}
+
+	return -EINVAL;
+}
+
+static int bcm2835_isp_node_try_fmt(struct file *file, void *priv,
+				    struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_fmt *fmt;
+
+	if (f->type != node->queue.type)
+		return -EINVAL;
+
+	if (!node_is_stats(node)) {
+		f->fmt.pix.width = max(min(f->fmt.pix.width, MAX_DIM),
+				       MIN_DIM);
+		f->fmt.pix.height = max(min(f->fmt.pix.height, MAX_DIM),
+					MIN_DIM);
+
+		fmt = find_format(f, node);
+		if (!fmt) {
+			f->fmt.pix.pixelformat =
+					get_default_format(node)->fourcc;
+			fmt = find_format(f, node);
+		}
```

Comment:
This is common to both paths (get_default_format should return V4L2_META_FMT_BCM2835_ISP_STATS on the stats node), so could be before the if.
Do the other fields need zero'ing on the stats node? For compressed formats width/height etc are 0, with only sizeimage being set. STATS is effectively equivalent to compressed data. (Almost query whether it should have the compressed flag set).

Reply:
If I understand your comment correctly, it would be a bit awkward to do this.   f->fmt is a union, and the .pix field is used for images and .meta is used for metadata.  If I were to combine the two paths, I will still end up with multiple if clauses (to setup the .pix.width and .pix.height before the find_format).

For the .meta field in the stats node (given it's a union) can only touch the .dataformat and .buffersize fields.

Reply:
Ah, I'd missed the union. 
You can still make it
```
fmt = find_format(f, node);
if (!fmt)
  fmt = get_default_format(node);
if (!node_is_stats(node)) {
  f->fmt.pix.pixelformat = fmt->fourcc;
 ...
} else {
 f->fmt.meta.dataformat = fmt->fourcc;
 ...
}
```
If the lookup succeeded then fmt is the correct format, otherwise you've chosen the default.
```

Reply:
The code would have to look something like:

```
if (!node_is_stats(node)) {
		f->fmt.pix.width = max(min(f->fmt.pix.width, MAX_DIM), MIN_DIM);
		f->fmt.pix.height = max(min(f->fmt.pix.height, MAX_DIM), MIN_DIM);
}

fmt = find_format(f, node);
if (!fmt)
  fmt = get_default_format(node);
if (!node_is_stats(node)) {
  f->fmt.pix.pixelformat = fmt->fourcc;
 ...
} else {
 f->fmt.meta.dataformat = fmt->fourcc;
 ...
}`
```
So we still have the extra if (!stats) block at the start.  However, I could simplify the original code by doing somthing like:

```
if (!stats) {
...
} else {
   f->fmt.meta.dataformat = V4L2_META_FMT_BCM2835_ISP_STATS;
   f->fmt.meta.buffersize = get_port_data(node)->minimum_buffer.size;
}

```

i.e. the stats clause does not need to go through find_format as we konw there is only going to be one stats format.

I'll use your judgement on what you think might be the better way.

Reply:
Done